File: Python/LeetCode/Trie/SplitString.py
from typing import List

class Solution:
    def partitionString(self, s: str) -> List[str]:
        res = []
        st = set()
        t = ""
        for c in s:
            t += c
            if t not in st:
                res.append(t)
                st.add(t)
                t = ""
        return res
    
    def partitionString2(self, s: str) -> List[str]:
        res = []
        cur = root = {}
        left = 0
        for i,c in enumerate(s):
            if c not in cur:
                cur[c] = {}
                res.append(s[left:i+1])
                left = i+1
                cur = root
            else:
                cur = cur[c]
        return res
#test
s = "abbccccd"
solution = Solution()
print(solution.partitionString(s))  # 输出 ["a","b","a","c","a","b","a"]

File: C_C++/LeetCode/Trie/SplitString.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<string> partitionString(string s)
    {
        vector<string> res;
        unordered_set<string> st;
        string t;
        for (char c : s)
        {
            t += c;
            if (st.insert(t).second)
            {
                res.push_back(t);
                t.clear();
            }
        }
        return res;
    }

    vector<string> partitionString2(string s)
    {
        struct Node
        {
            Node *son[26];
        };
        vector<string> res;
        int left = 0;
        Node *root = new Node();
        Node *cur = root;
        for (int i = 0; i < s.size();i++)
        {
            int c = s[i] - 'a';
            if(cur->son[c]==nullptr)
            {
                cur->son[c] = new Node();
                res.push_back(s.substr(left, i - left + 1));
                left = i+1;
                cur = root;
            }
            else
            {
                cur = cur->son[c];
            }
        }
        return res;
    }
};

int main()
{
    Solution s;
    string str = "abbccccd";
    vector<string> res = s.partitionString2(str);
    for (string &s : res)
        cout << s << endl;
}